spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/core/SpannerTemplate.java

```diff
@@ -219,87 +223,116 @@ public long executeDmlStatement(Statement statement) {
+		List<Mutation> mutations = this.mutationFactory.update(object, null);
+		emitSaveEventsApplyMutations(mutations, Collections.singletonList(object), null);
 	}
 
 	@Override
 	public void updateAll(Iterable objects) {
-		applyMutations(
-				getMutationsForMultipleObjects(objects,
-						(x) -> this.mutationFactory.update(x, null)));
+		List<Mutation> mutations = getMutationsForMultipleObjects(objects,
+				(x) -> this.mutationFactory.update(x, null));
+		emitSaveEventsApplyMutations(mutations, objects, null);
 	}
 
 	@Override
 	public void update(Object object, String... includeProperties) {
-		applyMutations(
-				this.mutationFactory.update(object,
-						(includeProperties.length == 0) ? null
-								: new HashSet<>(Arrays.asList(includeProperties))));
+		Set<String> incl = (includeProperties.length == 0) ? null
+				: new HashSet<>(Arrays.asList(includeProperties));
+		List<Mutation> mutations = this.mutationFactory.update(object, incl);
+		emitSaveEventsApplyMutations(mutations, Collections.singletonList(object), incl);
 	}
 
 	@Override
 	public void update(Object object, Set<String> includeProperties) {
-		applyMutations(this.mutationFactory.update(object, includeProperties));
+		List<Mutation> mutations = this.mutationFactory.update(object, includeProperties);
+		emitSaveEventsApplyMutations(mutations, Collections.singletonList(object), includeProperties);
 	}
 
 	@Override
 	public void upsert(Object object) {
-		applyMutations(this.mutationFactory.upsert(object, null));
+		List<Mutation> mutations = this.mutationFactory.upsert(object, null);
+		emitSaveEventsApplyMutations(mutations, Collections.singletonList(object), null);
 	}
 
 	@Override
 	public void upsertAll(Iterable objects) {
-		applyMutations(
+		List<Mutation> mutations =
 				getMutationsForMultipleObjects(objects,
-						(x) -> this.mutationFactory.upsert(x, null)));
+						(x) -> this.mutationFactory.upsert(x, null));
+		emitSaveEventsApplyMutations(mutations, objects, null);
 	}
 
 	@Override
 	public void upsert(Object object, String... includeProperties) {
-		applyMutations(
-				this.mutationFactory.upsert(object,
-						(includeProperties.length == 0) ? null
-								: new HashSet<>(Arrays.asList(includeProperties))));
+		Set<String> incl = (includeProperties.length == 0) ? null : new HashSet<>(Arrays.asList(includeProperties));
+		List<Mutation> mutations = this.mutationFactory.upsert(object, incl);
+		emitSaveEventsApplyMutations(mutations, Collections.singletonList(object), incl);
 	}
 
 	@Override
 	public void upsert(Object object, Set<String> includeProperties) {
-		applyMutations(this.mutationFactory.upsert(object, includeProperties));
+		List<Mutation> mutations = this.mutationFactory.upsert(object, includeProperties);
+		emitSaveEventsApplyMutations(mutations, Collections.singletonList(object), includeProperties);
+	}
+
+	private void emitSaveEventsApplyMutations(List<Mutation> mutations, Iterable entities,
```

Comment:
How about just `applySaveMutations`. The fact that it emits the save event, is an implementation detail.

Reply:
Renamed.